core/src/main/java/org/primefaces/extensions/component/keynote/KeynoteRenderer.java

```diff
@@ -0,0 +1,296 @@
+
+        // encodeCSS(context, "primefaces-extensions", "keynote/theme/" + keynote.getTheme() + ".css");
+        // encodeCSS(context, "primefaces", "primeicons/primeicons.css");
+        if (keynote.getTheme() != null && !"none".equals(keynote.getTheme())) {
+            encodeCSS(context, keynote.getLibrary(), keynote.getTheme());
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void encodeEnd(final FacesContext context, final UIComponent component) throws IOException {
+        final Keynote keynote = (Keynote) component;
+
+        encodeMarkup(context, keynote);
+        encodeScript(context, keynote);
+    }
+
+    /**
+     * Create the HTML markup for the DOM.
+     */
+    private void encodeMarkup(final FacesContext context, final Keynote keynote) throws IOException {
+        final ResponseWriter writer = context.getResponseWriter();
+        final String clientId = keynote.getClientId(context);
+        final String styleClass = getStyleClassBuilder(context)
+                    .add(CONTAINER_CLASS)
+                    .add(keynote.getStyleClass())
+                    .build();
+
+        writer.startElement("div", keynote);
+        writer.writeAttribute("id", clientId, null);
+        writer.writeAttribute(Attrs.CLASS, styleClass, "styleClass");
+        if (keynote.getStyle() != null) {
+            writer.writeAttribute(Attrs.STYLE, keynote.getStyle(), Attrs.STYLE);
+        }
+
+        writer.startElement("div", null);
+        writer.writeAttribute(Attrs.CLASS, SLIDES_CLASS, "class");
+
+        if (keynote.getVar() != null) {
+            // dynamic items
+            final Object value = keynote.getValue();
+            if (value != null) {
+                if (!(value instanceof Collection<?>)) {
+                    throw new FacesException("Value in Keynote must be of type Collection / List");
+                }
+
+                for (final UIComponent kid : keynote.getChildren()) {
+                    if (kid.isRendered() && !(kid instanceof UIKeynoteItem)) {
+                        // first render children like stamped elements, etc.
+                        renderChild(context, kid);
+                    }
+                }
+
+                final Collection<KeynoteItem> col = (Collection<KeynoteItem>) value;
+                for (final KeynoteItem keynoteItem : col) {
+                    // find ui item by type
+                    final UIKeynoteItem uiItem = keynote.getItem(keynoteItem.getType());
+
+                    if (uiItem.isRendered()) {
+                        // set data in request scope
+                        keynote.setData(keynoteItem);
+
+                        // render item
+                        renderItem(context, writer, keynote, uiItem);
+                    }
+                }
+            }
+        }
+        else {
+            // static items
+            for (final UIComponent kid : keynote.getChildren()) {
```

Comment:
Use index loop https://github.com/primefaces/primefaces/wiki/Contributing-to-PrimeFaces#performance-guidelines